clang/test/CIR/CodeGen/stmt-expr.c

```diff
@@ -0,0 +1,37 @@
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -fclangir-enable -emit-cir %s -o %t.cir
+// RUN: FileCheck --input-file=%t.cir %s
+
+// Yields void.
+void test1() { ({ }); }
+// CHECK: @test1
+//     CHECK: cir.scope {
+// CHECK-NOT:   cir.yield
+//     CHECK: }
+
+// Yields an l-value.
+void test2(int x) { ({ x;}); }
+// CHECK: @test2
+// CHECK: %{{.+}} = cir.scope {
+// CHECK:   %[[#V6:]] = cir.alloca !s32i, cir.ptr <!s32i>, ["tmp"]
+// CHECK:   %[[#V7:]] = cir.load %{{.+}} : cir.ptr <!s32i>, !s32i
+// CHECK:   cir.store %[[#V7]], %[[#V6]] : !s32i, cir.ptr <!s32i>
+// CHECK:   cir.yield %[[#V6]] : !cir.ptr<!s32i>
+// CHECK: }
```

Comment:
This seems wrong.

The scope yields the `#V6` pointer. However, `#V6` points to a memory space allocated within the scope, meaning it will be invalidated once the scope finishes.

To return the value from the scope, it should yield the value itself (e.g. `cir.yield %[[#V7]] : !s32i`), or copy the contents of the pointer to an allocation on the parent scope.

Reply:
That's true, I missed that, thanks for catching it. Ok, do you plan to make this change as part of this PR?

Reply:
@bcardosolopes patched! It's a bit hacky though: I'm just hoisting the alloca to the parent region.

I refactored some of the code to mimic what LLVM does: store the result on an alloca on the parent scope. Ideally, I would like the return value to be returned by `cir.scope`, but this would complicate other aspects of the pipeline, like generating copy constructors when lowering classes.

Reply:
This is fine, returning the value by `cir.scope` sounds like work for a transformation pass that can prove all necessary conditions to do it safely!